core/dbt/include/global_project/macros/materializations/models/table/columns_spec_ddl.sql

```diff
@@ -13,8 +13,8 @@
     {% for i in user_provided_columns %}
       {% set col = user_provided_columns[i] %}
       {% set constraints = col['constraints'] %}
-      {% set constraints_check = col['constraints_check'] %}
-      {{ col['name'] }} {{ col['data_type'] }} {% for x in constraints %} {{ x or "" }} {% endfor %} {% if constraints_check -%} check {{ constraints_check or "" }} {%- endif %} {{ "," if not loop.last }}
+      {{ col['name'] }} {{ col['data_type'] }} {% for x in constraints %} {{ "check" if x.type == "check" else "not null" if x.type == "not_null" else "unique" if x.type == "unique" else "primary key" if x.type == "primary_key" else "foreign key" if x.type == "foreign key" else ""
```

Comment:
Can we add a property to the constraint object solely for jinja templates so that this if block is calculated in python? I feel like it would be easier to maintain this across adapters in python than in jinja.

Reply:
I can see the argument for it, because this code (or similar) is repeated in several adapters. At the same time, I think the current approach allows for the greatest flexibility and support for syntax variations. What do you think @MichelleArk?

Reply:
The idea is that the python property on the `Constraint` object would represent the rendered out constraint, correct? Unfortunately the way constraints are rendered out is pretty adapter-specific (e.g. `primary key not enforced` for BigQuery, `primary key` for postgres). So we'd need this to be a method on the adapter - provided a `Constraint` object, render out its DDL. Most DDL generation is currently done in jinja already, but perhaps we could make this easier to maintain by splitting it out into a separate adapter.dispatch macro - perhaps `default__render_constraint(constraint)`?

Reply:
Thinking about this more, I do see another argument for this DDL rendering logic living in python as opposed to user-space (jinja). Although it's adapter-specific logic, there's really only one correct way to render constraint DDL given a `Constraint` object and it's very unlikely that individual dbt projects will want the ability to override that behaviour (maybe just to get a new constraint working faster than we can implement / accept contributions to the adapter python implementation). 

It feels similar to how the BigQuery adapter renders `PartitionConfig` via a `render` method: https://github.com/dbt-labs/dbt-bigquery/blob/main/dbt/adapters/bigquery/impl.py#L89-L97. The difference here being that it doesn't make sense as much sense to define the `Constraint` class in the `adapter` module given core requires it for parsing.

Reply:
So concretely what I'm suggesting here is:
* A new `@available.parse` method on [BaseAdapter](https://github.com/dbt-labs/dbt-core/blob/main/core/dbt/adapters/base/impl.py#L151), perhaps with the interface: `render_column_constraint(raw_column_constraint: Dict[str, str]) -> str`
* This new method would rely on an import of the `ColumnLevelConstraint` class from core to parse the `raw_constraint` to a `ColumnLevelConstraint` object, validate it, and render out the DDL. Unfortunately I don't see how we could implement a `render` method on the core-defined `ColumnLevelConstraint` class itself given the adapter-specific implementations.
* The adapter implementation would go from: 
```
{{ col['name'] }} {{ col['data_type'] }} {% for x in constraints %} {{ "check" if x.type == "check" else "not null" if x.type == "not_null" else "unique" if x.type == "unique" else "primary key" if x.type == "primary_key" else "foreign key" if x.type == "foreign key" else "" }} {{ x.expression or "" }} {% endfor %} {{ "," if not loop.last }}
```
to: 
```
{{ col['name'] }} {{ col['data_type'] }} {% for constraint in constraints %} {{ adapter.render_column_constraint(constraint) }}
```

Notes for dependent issues: 
* The warning logic that will end up using [`constraint_support`](https://github.com/dbt-labs/dbt-core/issues/7067) for raising warnings when users configure unsupported/unenforced constraints could live in this new method!
* We'd end up needing to implement a similar pattern for rendering [model-level constraints](https://github.com/dbt-labs/dbt-core/issues/6754) via a `render_model_constraint` method unless we can figure out a nice way to unify the two.

Reply:
^ This all tracks for me!

Rationale:
- This differs significantly by adapter, but _for each data platform_, there's really only one correct way to render a constraint string. There's no "business logic" here that can really differ by user of that platform.
- This ends up being a massive `if-else` engine, which is much more a strength of Python than of Jinja (which is nice for old-fashioned string templating, but that's about it). That's only truer with consideration for model-level constraints. We could also more easily write a unit test for this function, if it's written in Python :)
- The [`custom` constraint type](https://github.com/dbt-labs/dbt-core/pull/7130/files#diff-42cfc99b2fcd89ab75957b8999bee2b1af822f7d39953ca99f1f285c170557c1R151) should just render out its entire `expression` as a string. This is the escape hatch, in case a user can't get some wacky constraint string working with one of the more-strongly-typed options. dbt should assume that a `custom` constraint is supported and not enforced.

Reply:
@jtcohen6 @MichelleArk @mikealfare The latest changeset should implement the suggested changes. Let me know if you have any concerns!